source/extensions/tracers/datadog/agent_http_client.cc

```diff
@@ -0,0 +1,140 @@
+                                 TracerStats& stats)
+    : collector_cluster_(cluster_manager, cluster), cluster_(cluster),
+      reference_host_(reference_host), stats_(&stats) {}
+
+AgentHTTPClient::~AgentHTTPClient() {
+  for (const auto& [request_ptr, _] : handlers_) {
+    assert(request_ptr);
+    request_ptr->cancel();
+  }
+}
+
+// dd::HTTPClient
+
+dd::Expected<void> AgentHTTPClient::post(const URL& url, HeadersSetter set_headers,
+                                         std::string body, ResponseHandler on_response,
+                                         ErrorHandler on_error) {
+  ENVOY_LOG(debug, "flushing traces");
+
+  auto message = std::make_unique<Http::RequestMessageImpl>();
+  message->headers().setReferenceMethod(Http::Headers::get().MethodValues.Post);
+  message->headers().setReferencePath(url.path);
+  message->headers().setReferenceHost(reference_host_);
+
+  RequestHeaderWriter writer{message->headers()};
+  set_headers(writer);
+
+  message->body().add(body);
+  ENVOY_LOG(debug, "submitting trace(s) to {} with payload size {}", url.path, body.size());
+
+  if (!collector_cluster_.threadLocalCluster().has_value()) {
+    ENVOY_LOG(debug, "collector cluster '{}' does not exist", cluster_);
+    stats_->reports_skipped_no_cluster_.inc();
+    return dd::nullopt;
+  }
+
+  Http::AsyncClient::Request* request =
+      collector_cluster_.threadLocalCluster()->get().httpAsyncClient().send(
+          std::move(message), *this,
+          Http::AsyncClient::RequestOptions().setTimeout(std::chrono::milliseconds(1000)));
+  if (!request) {
+    stats_->reports_failed_.inc();
+    return dd::Error{dd::Error::ENVOY_HTTP_CLIENT_FAILURE, "Failed to create request."};
+  }
+
+  handlers_.emplace(request, Handlers{std::move(on_response), std::move(on_error)});
+  return dd::nullopt;
+}
+
+void AgentHTTPClient::drain(std::chrono::steady_clock::time_point) {}
+
+nlohmann::json AgentHTTPClient::config_json() const {
+  return nlohmann::json::object({
+      {"type", "Envoy::Extensions::Tracers::Datadog::AgentHTTPClient"},
+  });
+}
+
+// Http::AsyncClient::Callbacks
+
+void AgentHTTPClient::onSuccess(const Http::AsyncClient::Request& request,
+                                Http::ResponseMessagePtr&& response) {
+  const std::uint64_t status = Http::Utility::getResponseStatus(response->headers());
+  if (status != std::uint64_t(Http::Code::OK)) {
+    stats_->reports_dropped_.inc();
+  } else {
+    ENVOY_LOG(debug, "traces successfully submitted to datadog agent");
+    stats_->reports_sent_.inc();
+  }
+
+  auto found = handlers_.find(const_cast<Http::AsyncClient::Request*>(&request));
+  if (found == handlers_.end()) {
+    ENVOY_LOG(debug, "request at address 0x{} is not in the map of {} Handlers objects",
+              hex(reinterpret_cast<std::uintptr_t>(&request)), handlers_.size());
+    return;
+  }
+
+  auto& handlers = found->second;
```

Comment:
use a concrete type for this `auto`, as the type is not obvious from context.

https://google.github.io/styleguide/cppguide.html#Type_deduction

Reply:
done